include/date/solar_hijri.h

```diff
@@ -2198,16 +2202,46 @@ year_month_day::to_days() const NOEXCEPT
 
-    auto const y = static_cast<int>(y_);
+    auto const y = static_cast<int>(y_) - 475;
     auto const m = static_cast<unsigned>(m_);
     auto const d = static_cast<unsigned>(d_);
-    auto const era = static_cast<int>(((y-475) >= 0 ? (y-475) : (y-475)-(years_in_era-1)) / years_in_era);
+    auto const era_d = static_cast<int>(y >= 0 ? y : y-years_in_era+1) / static_cast<double>(years_in_era);
+    auto const era = static_cast<int>(era_d);
     auto const fdoe = static_cast<int>(epoch + era * days_in_era);
-    auto const yoe = static_cast<unsigned>(y - 475 - era * years_in_era);
-    auto const fdoyoe = static_cast<unsigned>(yoe*days_in_year);
+    auto const yoe = static_cast<unsigned>(y - era * years_in_era);
+
+    auto const years_in_first_cycle = static_cast<unsigned>(29);
+    auto const years_in_other_cycles = static_cast<unsigned>(33);
+    auto const years_in_period = static_cast<unsigned>(128);        // 29   + 3*33
+    auto const days_in_first_cycle = static_cast<unsigned>(10592);  // 28/4 + 29*365
+    auto const days_in_other_cycles = static_cast<unsigned>(12053); // 32/4 + 33*365
+    auto const days_in_period = static_cast<unsigned>(46751);       // days_in_first_cycle + 3*days_in_other_cycles;
+    auto const period_d = static_cast<double>(yoe/years_in_period);
+    auto const period = static_cast<unsigned>(period_d);
+    auto const yop = yoe%years_in_period;
+    auto const fdop = period*days_in_period;
+    auto const cycle = yop < 29 ? 0 : static_cast<unsigned>((yop-29)/years_in_other_cycles + 1);
+    auto const yoc = yop < 29 ? yop : (yop-29)%years_in_other_cycles;
+    auto const fdoc = cycle > 0 ? days_in_first_cycle + (cycle-1)*days_in_other_cycles : 0;
+    auto const group = yoc < 1 ? 0 : static_cast<unsigned>((yoc-1) / 4);
+    auto const yog = static_cast<int>(yoc < 1 ? -1 : (yoc-1) % 4); // yoc < 1 ? -1 : (yoc-1) % 4;
+    auto const fdoyog = group*1461 + (yog+1)*365;
+    auto const fdoyoe = fdop + fdoc + fdoyog;
+
     auto const doy = 30*(m-1) + ((m > 6) ? 6 : m-1) + d-1;                     // [0, 365]
     auto const doe = fdoe + fdoyoe + doy;
     return days{doe - unix_time_shift};
+    // auto const era_d = static_cast<int>(y >= 0 ? y : y-years_in_era+1) / static_cast<double>(years_in_era);
+    // auto const era = static_cast<int>(era_d);
+    // // auto const era = static_cast<int>(((y-475) >= 0 ? (y-475) : (y-475)-(years_in_era-1)) / years_in_era);
+    // auto const fdoe = static_cast<int>(epoch + era * days_in_era);
+    // auto const yoe = static_cast<unsigned>(y - era * years_in_era);
+    // auto const fdoyoe = static_cast<double>(yoe*days_in_year);
+    // // auto const fdoyoe = static_cast<unsigned>(yoe*days_in_year);
+    // auto const fdoyoe_i = static_cast<int>(fdoyoe);
+    // auto const doy = 30*(m-1) + ((m > 6) ? 6 : m-1) + d-1;                     // [0, 365]
+    // auto const doe = fdoe + fdoyoe_i + doy;
+    // return days{doe - unix_time_shift};
```

Comment:
Once you've achieved confidence in this new algorithm, can you remove the old one before we commit to master?

Reply:
Done!